### PR TITLE
syncer: handle connetion refused error

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
### What problem does this PR solve? 

After kill upstream mysql, will still show `running` in `query-status`'s result, because `GetEvent` gets `context.DeadlineExceeded` error and will retry forever.

### What is changed and how it works?

1. add `MaxReconnectAttempts` for `BinlogSyncerConfig`, and binlog replication  in go-mysql will return error(connection refused) after retry some times.( if `MaxReconnectAttempts` is 0, go-mysql will retry forever)
2. handle `connection refused` error when `GetEvent` failed.

### Check List 

Tests 

 - Unit test
 - Manual test (kill upstream mysql, and check status)

Related changes

 - Need to be included in the release note

### CODE REVIEW

File: syncer/syncer.go [hunk 1]
```diff
@@ -2430,6 +2436,7 @@ func (s *Syncer) setSyncCfg() {
 		UseDecimal:              true,
 		VerifyChecksum:          true,
 		TimestampStringLocation: s.timezone,
+		MaxReconnectAttempts:    maxBinlogSyncerReconnect,
```
Comment (csuzhangxc) [inline syncer/syncer.go]: How about giving more config items like `ReadTimeout`, `HeartbeatPeriod`, etc. (also see what we did in relay, https://github.com/pingcap/dm/blob/1724a39f692145faaec7e4f878b1ec622bbd0a74/relay/relay.go#L120).

File: syncer/syncer.go [hunk 2]
```diff
@@ -1166,6 +1169,9 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 				return err1
 			}
 			continue
+		} else if isConnectionRefusedError(err) {
+			tctx.L().Error("fail to get binlog event", zap.Error(err))
+			return err
```
[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply syncer/syncer.go]: What's the reason？

[the remark this answers is not in the document]
Comment (csuzhangxc) [reply syncer/syncer.go]: If we don't ignore/retry for the `err`, then we will return it (and we have do that in `if err != nil {` block from l1177.)

File: relay/relay.go [hunk 3]
```diff
@@ -118,19 +117,15 @@ type Relay struct {
...
-		Host:             cfg.From.Host,
-		Port:             uint16(cfg.From.Port),
-		User:             cfg.From.User,
-		Password:         cfg.From.Password,
-		Charset:          cfg.Charset,
-		UseDecimal:       true, // must set true. ref: https://github.com/pingcap/tidb-enterprise-tools/pull/272
-		ReadTimeout:      slaveReadTimeout,
-		HeartbeatPeriod:  masterHeartbeatPeriod,
-		VerifyChecksum:   true,
-		DisableRetrySync: true, // the retry of go-mysql has some problem now, we disable it and do the retry in relay first.
```
Comment (WangXiangUSTC) [inline relay/relay.go:132]: this issue is already fixed

File: syncer/syncer.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply syncer/syncer.go]: update in f2ca52f

File: syncer/syncer.go (hunk 1 above)
Comment (WangXiangUSTC) [reply syncer/syncer.go]: update in f2ca52f

File: pkg/binlog/common/replication.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	"time"
+
+	"github.com/siddontang/go-mysql/replication"
+)
+
+var (
+	// max reconnection times for binlog syncer in go-mysql
+	maxBinlogSyncerReconnect = 60
+
+	slaveReadTimeout      = 1 * time.Minute  // slave read binlog data timeout, ref: https://dev.mysql.com/doc/refman/8.0/en/replication-options-slave.html#sysvar_slave_net_timeout
```
Comment (csuzhangxc) [inline pkg/binlog/common/replication.go:26]: Maybe we can use `slaveReadTimeout` to replace `eventTimeout` in _syncer/streamer_controller.go_ and _relay/reader/reader.go_?

File: pkg/binlog/common/replication.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+	masterHeartbeatPeriod = 30 * time.Second // master server send heartbeat period: ref: `MASTER_HEARTBEAT_PERIOD` in https://dev.mysql.com/doc/refman/8.0/en/change-master-to.html
```
Comment (csuzhangxc) [inline pkg/binlog/common/replication.go:27]: It seems we should add a `TODO` in some place to indicate we should add _heartbeat event_ support for _local_ reader later.

File: syncer/syncer.go [hunk 6]
```diff
@@ -1188,16 +1188,6 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 			return nil
 		} else if err == context.DeadlineExceeded {
 			s.tctx.L().Info("deadline exceeded when fetching binlog event")
-			eventTimeoutCounter += eventTimeout
-			if eventTimeoutCounter < maxEventTimeout {
```
[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply syncer/syncer.go:1192]: removed

File: pkg/binlog/common/replication.go (hunk 5 above)
Comment (WangXiangUSTC) [reply pkg/binlog/common/replication.go:27]: add todo in 5f79104

File: pkg/binlog/common/replication.go (hunk 4 above)
Comment (WangXiangUSTC) [reply pkg/binlog/common/replication.go]: replaced in 5f79104

**Comment by WangXiangUSTC** [comment]
@lichunzhu @csuzhangxc PTAL again

**Pull request merged**